### ISSUE TITLE
newtype

### ISSUE THREAD

**Issue by matthewleon** [paf31/purescript-lists-fast#1]
Probably worth making a `newtype` so that polymorphic functions take advantage of the fast `map`, no?

**Comment by paf31** [comment]
Sounds like a good idea. Maybe it can implement `Filterable` and `Applicative` as well.

**Comment by matthewleon** [comment]
What would change for `Applicative`?

**Comment by paf31** [comment]
I was referring to the zippy applicative instance.

**Comment by matthewleon** [comment]
My inclination is to, by default, have the same `Apply` semantics as `List`. Is there a reason for changing this? Perhaps a Haskell-style `ZipList` is called for?

**Comment by matthewleon** [comment]
Or rather, a separate `newtype` that works like these: https://pursuit.purescript.org/packages/purescript-lists/4.8.0/docs/Data.List.ZipList

**Comment by matthewleon** [comment]
A tricky thing I'm running into with the `newtype`: reexporting versions of all the original `List` functions that work on the `FastList` newtype sometimes means stuff like this:

```
tail :: forall a. FastList a -> Maybe (FastList a)
tail (FastList xs) = FastList <$> L.tail xs
```

I'm guessing the compiler output for this will technically be slower than the `List` implementation, because it will add at least one function invocation. Which makes some functions a tiny bit slower on `FastList` than on `List`.

On the other hand, rewriting these functions for `FastList` seems like a total drag.

Any advice?

**Comment by paf31** [comment]
I don't think we should reexport everything for the newtype, only those things which can be switched out conveniently via a type class.
